src/v/cloud_storage/remote_partition.cc

```diff
@@ -800,4 +801,88 @@ remote_partition::iterator remote_partition::upper_bound(kafka::offset o) {
 }
 
+/**
+ * The caller is responsible for determining whether it is appropriate
+ * to delete data in S3, for example it is not appropriate if this is
+ * a read replica topic.
+ *
+ * Q: Why is erase() part of remote_partition, when in general writes
+ *    flow through the archiver and remote_partition is mainly for
+ *    reads?
+ * A: Because the erase operation works in terms of what it reads from
+ *    S3, not the state of the archival metadata stm (which can be out
+ *    of date if e.g. we were not the leader)
+ */
+ss::future<> remote_partition::erase() {
+    // TODO: Edge case 1
+    // There is a rare race in which objects might get left behind in S3.
+    //
+    // Consider 3 nodes 1,2,3.
+    // Node 1 is controller leader.  Node 3 is this partition's leader.
+    // Node 1 receives a delete topic request, replicates to 2, and
+    // acks to the client.
+    // Nodes 1 and 2 proceed to tear down the partition.
+    // Node 3 experiences a delay seeing the controller message,
+    // and is still running archival fiber, writing new objects to S3.
+    // Now stop Node 3, and force-decommission it.
+    // The last couple of objects written by Node 3 are left behind in S3.
+    // (TODO: perhaps document this as a caveat of force-decom operations?)
+
+    // TODO: Edge case 2
+    // Archiver writes objects to S3 before it writes an update manifest.
+    // If a topic is deleted while undergoing writes, it may end up leaving
+    // some garbage segments behind.
+    // (TODO: fix this by implementing a scrub operation that finds objects
+    //  not mentioned in the manifest, and call this before erase())
+
+    static ss::lowres_clock::duration erase_timeout = 60s;
+    static ss::lowres_clock::duration erase_backoff = 1s;
+    retry_chain_node local_rtc(erase_timeout, erase_backoff, &_rtc);
+
+    // Read the partition manifest fresh: we might already have
+    // dropped local archival_stm state related to this partition.
+    partition_manifest manifest(
+      _manifest.get_ntp(), _manifest.get_revision_id());
+
+    auto manifest_path = manifest.get_manifest_path();
+    auto manifest_get_result = co_await _api.maybe_download_manifest(
+      _bucket, manifest_path, manifest, local_rtc);
+
+    if (manifest_get_result != download_result::notfound) {
+        // Erase all segments
+        for (const auto& i : manifest) {
+            auto path = manifest.generate_segment_path(i.first, i.second);
+            vlog(_ctxlog.debug, "Erasing segment {}", path);
+            // On failure, we throw: this should cause controller to retry
+            // the topic deletion operation that called us, until it eventually
+            // succeeds.
+            // TODO: how do we get out of this state if e.g. the bucket no
+            // longer exists, or the S3 backend is just permanently gone?
+            // TODO: S3 API has a plural delete API, which would be more
+            // suitable.
+            co_await _api.delete_object(
```

Comment:
the return value of `delete_object` is ignored

Reply:
Good spot, I've added error handling to these calls: treating timeouts as retryable  and other errors as a cue to give up (explaining why in comments)